skbio/alignment/_path.py

```diff
@@ -237,29 +265,54 @@ def from_indices(cls, indices, gap=-1):
             Each element in the array is the index in the corresponding sequence.
         gap : int or "mask", optional
             The value which represents a gap in the alignment. Defaults to -1, but
-            can be other values. If "mask", `indices` must be a masked array. Cannot
-            use "del".
+            can be other values. If "mask", ``indices`` must be an
+            ``np.ma.MaskedArray``. Cannot use "del".
 
         Returns
         -------
         AlignPath
             The alignment path created from the given indices.
+
+        Notes
+        -----
+        The returned alignment path will span across the entire tabular MSA. Its start
+        positions will be uniformly zeros.
+
         """
         if gap == "mask":
-            return cls.from_bits(np.ma.getmask(indices))
+            return cls.from_bits(
+                np.ma.getmask(indices),
+                [x[np.argmax(x != indices.fill_value)] for x in indices],
+            )
         else:
-            return cls.from_bits(indices == gap)
+            indices = np.asarray(indices)
+            return cls.from_bits(
+                indices == gap, [x[np.argmax(x != gap)] for x in indices]
```

Comment:
@mataton After some exploration, the best non-Cython solution I learned is:

Replace:

```python
[x[np.argmax(x != gap)] for x in indices]
```

With:

```python
indices[np.arange(indices.shape[0]), np.argmax(indices != gap, axis=1)]
```

This [thread](https://stackoverflow.com/questions/16243955/numpy-first-occurrence-of-value-greater-than-existing-value) has some nice discussions and benchmarks. The most interesting thing is Mr Fooz's reply.